packages/core/contracts/financial-templates/common/FundingRateApplier.sol

```diff
@@ -77,13 +89,60 @@ abstract contract FundingRateApplier is Testable, Lockable {
      * @notice Constructs the FundingRateApplier contract. Called by child contracts.
      * @param _finderAddress Finder used to discover financial-product-related contracts.
      */
-    constructor(address _finderAddress) public {
-        finder = FinderInterface(_finderAddress);
+    constructor(
+        FixedPoint.Unsigned memory _fundingRateBondPercentage,
+        FixedPoint.Unsigned memory _fundingRateRewardRate,
+        bytes32 _fundingRateIdentifier,
+        address _collateralAddress,
+        address _finderAddress,
+        address _timerAddress
+    ) public FeePayer(_collateralAddress, _finderAddress, _timerAddress) {
+        fundingRate.updateTime = getCurrentTime();
+
+        // Seed the cumulative multiplier as 1, from which it will be scaled as funding rates are applied over time.
+        fundingRate.cumulativeMultiplier = FixedPoint.fromUnscaledUint(1);
         emergencyShutdownTimestamp = 0;
-        lastUpdateTime = getCurrentTime();
 
-        // Seed the initial funding rate in the cumulativeFundingRateMultiplier 1.
-        cumulativeFundingRateMultiplier = FixedPoint.fromUnscaledUint(1);
+        fundingRate.rewardRate = _fundingRateRewardRate;
+        fundingRate.identifier = _fundingRateIdentifier;
+        fundingRate.bondPercentage = _fundingRateBondPercentage;
+    }
+
+    function proposeNewRate(FixedPoint.Signed memory rate, uint256 timestamp)
+        external
+        nonReentrant()
+        updateFundingRate()
+        returns (FixedPoint.Unsigned memory totalBond)
+    {
+        require(fundingRate.proposalTime == 0, "Proposal in progress");
+
+        // Timestamp must be after the last funding rate update time, within the last 30 minutes, and it cannot be more
+        // than 90 seconds ahead of the block timestamp.
+        uint256 currentTime = getCurrentTime();
+        uint256 updateTime = fundingRate.updateTime;
+        require(
+            timestamp > updateTime && timestamp >= currentTime.sub(30 minutes) && timestamp <= currentTime.add(90),
+            "Invalid proposal time"
+        );
+
+        // Set the proposal time in order to allow this contract to track this request.
+        fundingRate.proposalTime = timestamp;
+
+        OptimisticOracle optimisticOracle = _getOptimisticOracle();
+
+        // Set up optmistic oracle.
+        bytes32 identifier = fundingRate.identifier;
+        optimisticOracle.requestPrice(identifier, timestamp, collateralCurrency, 0);
```

Comment:
Why do you set reward to 0 here, and then you pay out the reward in `_getLatestFundingRate()`? Could we just remove the rewards param from this `optimisticOracle` method?

Reply:
As discussed IRL, the optimistic oracle is meant to be more general, so even though we don't use the rewards, it's likely that others might want to.